File: apps/macos/Sources/PolyglanceKit/TranslationHistoryStore.swift
import Foundation

public struct TranslationRecord: Identifiable, Codable, Equatable, Sendable {
    public var id: UUID
    public var timestamp: Date
    public var sourceText: String
    public var targetText: String
    public var sourceLang: String
    public var targetLang: String
    public var provider: String
    public var isFavorite: Bool

    public init(
        id: UUID = UUID(),
        timestamp: Date = Date(),
        sourceText: String,
        targetText: String,
        sourceLang: String,
        targetLang: String,
        provider: String,
        isFavorite: Bool = false
    ) {
        self.id = id
        self.timestamp = timestamp
        self.sourceText = sourceText
        self.targetText = targetText
        self.sourceLang = sourceLang
        self.targetLang = targetLang
        self.provider = provider
        self.isFavorite = isFavorite
    }
}

@MainActor
public final class TranslationHistoryStore: ObservableObject {
    @Published public private(set) var records: [TranslationRecord] = []

    private let fileURL: URL

    public static let shared = TranslationHistoryStore()

    public init(fileURL: URL? = nil) {
        if let fileURL {
            self.fileURL = fileURL
        } else {
            let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            let dir = appSupport.appendingPathComponent("Polyglance", isDirectory: true)
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            self.fileURL = dir.appendingPathComponent("translation_history.json")
        }
        load()
    }

    public func addRecord(
        sourceText: String,
        targetText: String,
        sourceLang: String,
        targetLang: String,
        provider: String
    ) {
        let trimmedSource = sourceText.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedTarget = targetText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedSource.isEmpty, !trimmedTarget.isEmpty else { return }

        // Remove identical duplicate if at the very top
        if let first = records.first, first.sourceText == trimmedSource && first.targetText == trimmedTarget {
            return
        }

        let record = TranslationRecord(
            sourceText: trimmedSource,
            targetText: trimmedTarget,
            sourceLang: sourceLang,
            targetLang: targetLang,
            provider: provider
        )
        records.insert(record, at: 0)
        if records.count > 100 {
            records = Array(records.prefix(100))
        }
        save()
    }

    public func toggleFavorite(id: UUID) {
        if let idx = records.firstIndex(where: { $0.id == id }) {
            records[idx].isFavorite.toggle()
            save()
        }
    }

    public func toggleFavoriteForCurrent(sourceText: String, targetText: String) {
        let trimmed = sourceText.trimmingCharacters(in: .whitespacesAndNewlines)
        if let idx = records.firstIndex(where: { $0.sourceText == trimmed }) {
            records[idx].isFavorite.toggle()
            save()
        } else if !trimmed.isEmpty && !targetText.isEmpty {
            let record = TranslationRecord(
                sourceText: trimmed,
                targetText: targetText,
                sourceLang: "",
                targetLang: "",
                provider: "",
                isFavorite: true
            )
            records.insert(record, at: 0)
            save()
        }
    }

    public func isFavorite(sourceText: String) -> Bool {
        let trimmed = sourceText.trimmingCharacters(in: .whitespacesAndNewlines)
        return records.first(where: { $0.sourceText == trimmed })?.isFavorite ?? false
    }

    public func deleteRecord(id: UUID) {
        records.removeAll { $0.id == id }
        save()
    }

    public func clearAll() {
        records.removeAll()
        save()
    }

    private func load() {
        guard let data = try? Data(contentsOf: fileURL),
              let list = try? JSONDecoder().decode([TranslationRecord].self, from: data) else {
            return
        }
        self.records = list
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(records) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }
}

File: apps/windows/src/Polyglance.Core/Services/TranslationHistoryStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Polyglance.Core.Services;

public sealed class TranslationRecord
{
    [JsonPropertyName("id")]
    public Guid Id { get; set; } = Guid.NewGuid();

    [JsonPropertyName("timestamp")]
    public DateTimeOffset Timestamp { get; set; } = DateTimeOffset.UtcNow;

    [JsonPropertyName("source_text")]
    public string SourceText { get; set; } = "";

    [JsonPropertyName("target_text")]
    public string TargetText { get; set; } = "";

    [JsonPropertyName("source_lang")]
    public string SourceLang { get; set; } = "";

    [JsonPropertyName("target_lang")]
    public string TargetLang { get; set; } = "";

    [JsonPropertyName("provider")]
    public string Provider { get; set; } = "";

    [JsonPropertyName("is_favorite")]
    public bool IsFavorite { get; set; }
}

public sealed class TranslationHistoryStore
{
    private static readonly Lazy<TranslationHistoryStore> _shared = new(() => new TranslationHistoryStore());
    public static TranslationHistoryStore Shared => _shared.Value;

    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };
    private readonly string _filePath;
    private readonly object _lock = new();
    private List<TranslationRecord> _records = new();

    public TranslationHistoryStore(string? filePath = null)
    {
        if (!string.IsNullOrWhiteSpace(filePath))
        {
            _filePath = filePath;
        }
        else
        {
            string appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
            _filePath = Path.Combine(appData, "Polyglance", "translation_history.json");
        }

        Load();
    }

    public IReadOnlyList<TranslationRecord> GetRecords()
    {
        lock (_lock)
        {
            return _records.ToList();
        }
    }

    public IReadOnlyList<TranslationRecord> GetFavorites()
    {
        lock (_lock)
        {
            return _records.Where(r => r.IsFavorite).ToList();
        }
    }

    public void AddRecord(string sourceText, string targetText, string sourceLang, string targetLang, string provider)
    {
        string trimmedSource = (sourceText ?? "").Trim();
        string trimmedTarget = (targetText ?? "").Trim();
        if (string.IsNullOrEmpty(trimmedSource) || string.IsNullOrEmpty(trimmedTarget))
            return;

        lock (_lock)
        {
            if (_records.Count > 0 && _records[0].SourceText == trimmedSource && _records[0].TargetText == trimmedTarget)
                return;

            var record = new TranslationRecord
            {
                Id = Guid.NewGuid(),
                Timestamp = DateTimeOffset.UtcNow,
                SourceText = trimmedSource,
                TargetText = trimmedTarget,
                SourceLang = sourceLang ?? "",
                TargetLang = targetLang ?? "",
                Provider = provider ?? "",
                IsFavorite = false
            };

            _records.Insert(0, record);
            if (_records.Count > 100)
            {
                _records = _records.Take(100).ToList();
            }

            Save();
        }
    }

    public void ToggleFavorite(Guid id)
    {
        lock (_lock)
        {
            var item = _records.FirstOrDefault(r => r.Id == id);
            if (item != null)
            {
                item.IsFavorite = !item.IsFavorite;
                Save();
            }
        }
    }

    public void ToggleFavoriteForCurrent(string sourceText, string targetText)
    {
        string trimmed = (sourceText ?? "").Trim();
        if (string.IsNullOrEmpty(trimmed))
            return;

        lock (_lock)
        {
            var item = _records.FirstOrDefault(r => r.SourceText == trimmed);
            if (item != null)
            {
                item.IsFavorite = !item.IsFavorite;
                Save();
            }
            else if (!string.IsNullOrEmpty(targetText?.Trim()))
            {
                var record = new TranslationRecord
                {
                    Id = Guid.NewGuid(),
                    Timestamp = DateTimeOffset.UtcNow,
                    SourceText = trimmed,
                    TargetText = targetText.Trim(),
                    SourceLang = "",
                    TargetLang = "",
                    Provider = "",
                    IsFavorite = true
                };
                _records.Insert(0, record);
                Save();
            }
        }
    }

    public bool IsFavorite(string sourceText)
    {
        string trimmed = (sourceText ?? "").Trim();
        if (string.IsNullOrEmpty(trimmed))
            return false;

        lock (_lock)
        {
            return _records.FirstOrDefault(r => r.SourceText == trimmed)?.IsFavorite ?? false;
        }
    }

    public void DeleteRecord(Guid id)
    {
        lock (_lock)
        {
            _records.RemoveAll(r => r.Id == id);
            Save();
        }
    }

    public void ClearAll()
    {
        lock (_lock)
        {
            _records.Clear();
            Save();
        }
    }

    private void Load()
    {
        lock (_lock)
        {
            try
            {
                if (File.Exists(_filePath))
                {
                    string json = File.ReadAllText(_filePath);
                    _records = JsonSerializer.Deserialize<List<TranslationRecord>>(json) ?? new();
                }
            }
            catch
            {
                _records = new();
            }
        }
    }

    private void Save()
    {
        try
        {
            string? dir = Path.GetDirectoryName(_filePath);
            if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
            {
                Directory.CreateDirectory(dir);
            }

            string json = JsonSerializer.Serialize(_records, JsonOptions);
            File.WriteAllText(_filePath, json);
        }
        catch
        {
            // Ignore persistence errors
        }
    }
}
